src/sagemaker_containers/training.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import importlib
+import os
+import shlex
+import subprocess
+
+import sagemaker_containers.training_environment as env
+
+_PYTHONPATH = 'PYTHONPATH'
+
+
+class Training(object):
+    """Training class to start training in the container.
+
+    When training starts, the `Training` object ensures that the user script (or Python package)
+    is imported and executed in a different process, reporting success in the end of the training.
+
+    `Training` is responsible for error handling and metrics reporting and reports failures in case
+    of exceptions.
+    """
+
+    def __init__(self, training_process_fn):
+        """Constructs a `Training` instance.
+
+        When a training starts, e.g. `Training(training_process_fn).run()`, the `Training` instance
+        will execute the following steps:
+
+            1 - install required Python dependencies
+            2 - start required customs metrics
+            3 - load training environment
+            4 - download the user script (or Python package) containing the functions required by the framework
+            5 - start training process by invoking `training_process_fn`
+            6 - report success/failure
+
+        Args:
+            training_process_fn: a training process function that will be executed by the `Training`. Signature:
+
+                * Args:
+                    * `training_environment`(TrainingEnvironment): an instance of the `TrainingEnvironment`.
+        """
+        self.environment = env.TrainingEnvironment()
+        self.training_process_fn = training_process_fn
+
+    # TODO(mvsusp) - add update_training_environment_fn to allow frameworks to update the training env.
```

Comment:
what does update mean? user can just update/attributes to the instance without adding a special function (b/c python)

Reply:
for the following scenario:

```python
    
def framework_train_fn(user_module, training_env):
    model = user_module.train(training_env.channel_input_dirs, training_env.hyperparameters)
    save_model(model, model_file)

def update_training_environment_fn(training_env):
    training_env.world_size = len(training_env.hosts * training_env.num_gpus)
    training_env.rank = 2
    return training_env


Training.from_train_fn(train_fn=framework_train_fn,  
                                      environment_fn=update_training_environment_fn)
```

Put it is outside the scope this PR. We can discuss implementation details later.